### PR TITLE
[INTEL_MKL] Windows OpenMP support

### PR DESCRIPTION
**Pull request by bani-intelaipg**
Enable Windows OpenMP support for Intel-MKL to improve performance on CPUs

### CODE REVIEW

File: third_party/llvm_openmp/openmp.bzl [hunk 1]
```diff
@@ -19,3 +30,72 @@ def dict_add(*dictionaries):
...
+def select_os_specific(L, M, W):
+    return select({
+        "@org_tensorflow//tensorflow:linux_x86_64": L,
+        "@org_tensorflow//tensorflow:macos": M,
+        "@org_tensorflow//tensorflow:windows": W,
+        "//conditions:default": L,
+    })
+
+def libname_os_specific():
+    return "" + select_os_specific(L = "libiomp5.so", M = "libiomp5.dylib", W = "libiomp5md.dll")
```
Comment (penpornk) [inline third_party/llvm_openmp/openmp.bzl:43]: Nit: We can refactor "libiomp5" out.
```suggestion
    return "libiomp5" + select_os_specific(L = ".so", M = ".dylib", W = ".dll")
```

File: third_party/llvm_openmp/openmp.bzl [hunk 2]
```diff
@@ -19,3 +30,72 @@ def dict_add(*dictionaries):
...
+# ......
+# ......
+
+# MacOS build has not been tested, however since the MacOS build of openmp
+# uses the same configuration as Linux, the following should work.
+def libiomp5_cc_binary(libname, cppsources, srcdeps, common_includes):
+    cc_binary(
+        name = libname,
+        srcs = cppsources + srcdeps +
+            select_os_specific(
```
Comment (penpornk) [inline third_party/llvm_openmp/openmp.bzl:63]: Please create another rule `if_linux_or_macos(T, F)` that returns `T` when it's linux or macos, then use this clause for everything that linux and macos use the same values.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

**Review by penpornk — APPROVED** [review]
Thank you for the fix!

**Comment by bani-intelaipg** [comment]
@penpornk bazel throes some error: Error in cc_binary: cc_binary 'name' attribute must be a string. Is there a suggested fix for this?
`ERROR: Traceback (most recent call last):
        File "/localdisk/bani/inteltf1-build/3de45b8df3514ee70c6c7e96234456ef/external/llvm_openmp/BUILD.bazel", line 226, column 10, in <tople
                cc_binary(
        File "/localdisk/bani/inteltf1-build/3de45b8df3514ee70c6c7e96234456ef/external/rules_cc/cc/defs.bzl", line 39, column 21, in cc_binary
                native.cc_binary(**_add_tags(attrs))
Error in cc_binary: cc_binary 'name' attribute must be a string
`

**Comment by penpornk** [comment]
@bani-intelaipg Ah, we probably need to generate the rule like you did then. (Sorry about that. I'm not a bazel expert either.)
Can we do this? Define a genrule in `openmp.bzl` and call it from the `BUILD` file:
openmp.bzl:
```
def tf_gen_libiomp(name, sources, common_includes):
    # Linux and MacOS
    for ext in [".so", ".dylib"]:
        cc_binary(
            name = name + ext,
            srcs = sources + [
                "runtime/src/z_Linux_util.cpp",
                "runtime/src/kmp_gsupport.cpp",
                "runtime/src/z_Linux_asm.S",
            ],
            additional_linker_inputs = ["ldscript"],
            copts = ["-Domp_EXPORTS -D_GNU_SOURCE -D_REENTRANT"],
            includes = common_includes,
            linkopts = ["-lpthread -ldl -Wl,--version-script=$(location :ldscript)"],
            linkshared = True,
            visibility = ["//visibility:public"],
        )

    # Windows
    cc_binary(
        name = name + ".dll",
        srcs = sources + [
            "runtime/src/z_Windows_NT_util.cpp",
            "runtime/src/z_Windows_NT-586_util.cpp",
            ":openmp_asm",
        ],
        additional_linker_inputs = [":generate_def"],
        copts = ["/Domp_EXPORTS /D_M_AMD64 /DOMPT_SUPPORT=0 /D_WINDOWS /D_WINNT /D_USRDLL"],
        includes = common_includes,
        linkopts = ["/MACHINE:X64"],
        linkshared = True,
        visibility = ["//visibility:public"],
        win_def_file = ":generate_def",
    )
```

BUILD file:
```
tf_gen_libiomp(
    name = "libiomp5",
    common_includes = common_includes,
    sources = cppsources + srcdeps + headers,
)
```

If this doesn't work, please feel free to go back to your original changes which you have tested (since we are short on time) but use the `select_os_specific_2` to avoid duplicate strings between linux and mac.

**Comment by bani-intelaipg** [comment]
We are reverting as you suggested. Will push the commit soon.

**Comment by penpornk** [comment]
Thank you! 
Does your original commit build for you? I tried building the `libiomp5.so` target on Ubuntu and got this error:
```
ERROR: /root/windows/tensorflow/third_party/llvm_openmp/BUILD:129:18: in genrule rule //third_party/llvm_openmp:config_omp: target '//third_party/llvm:expand_cmake_vars' is not visible from target '//third_party/llvm_openmp:config_omp'. Check the visibility declaration of the former target if you think the dependency is legitimate
ERROR: Analysis of target '//third_party/llvm_openmp:libiomp5.so' failed; build aborted: Analysis of target '//third_party/llvm_openmp:config_omp' failed
```

**Comment by bani-intelaipg** [comment]
Hi @penpornk after reverting the changes the build passes in both Linux and Windows.

**Review by penpornk — APPROVED** [review]
Thank you for the changes! Let's see if other tests are okay.

**Review by gbaned — APPROVED** [review]

**Comment by bani-intelaipg** [comment]
@penpornk Please let us know if anything remains to be done before including this PR.

**Comment by penpornk** [comment]
@bani-intelaipg I have pulled this PR in yesterday and fixed additional issues from internal tests (internally) this morning. It is not merged yet because of some infra issues. I'm not sure if it will get merged in time, but I'll keep trying the merge.

**Comment by penpornk** [comment]
@bani-intelaipg Seems like the merge just went through. The PR should be updated shortly. :)

**Pull request merged**